### PR TITLE
Rule S5693: Allowing requests with excessive content length is security-sensitive

### LINKED ISSUE #3980: Rule S5693: Allowing requests with excessive content length is security-sensitive
https://jira.sonarsource.com/browse/MMF-2159
https://jira.sonarsource.com/browse/RSPEC-5693
SonarSource/security-expected-issues#240

### PR DESCRIPTION
**Pull request by csaba-sagi-sonarsource**
Fixes #3980

### CODE REVIEW

File: analyzers/rspec/vbnet/S5693_vb.net.html [hunk 1]
```diff
@@ -0,0 +1,28 @@
+<p>Rejecting requests with significant content length is a good practice to control the network traffic intensity and thus resource consumption in
```
Comment (costin-zaharia-sonarsource) [inline analyzers/rspec/vbnet/S5693_vb.net.html:1]: The rule was specified for C# but not yet for VB.NET. We need to update the jira rspec by adding a subtask for VB with code examples.

File: analyzers/src/SonarAnalyzer.Common/Rules/BypassingAccessibilityBase.cs [hunk 2]
```diff
@@ -26,16 +28,22 @@ public abstract class BypassingAccessibilityBase<TSyntaxKind> : SonarDiagnosticA
...
+        protected FieldAccessTracker<TSyntaxKind> FieldAccessTracker { get; set; }
+
+        protected DiagnosticDescriptor Rule { get; }
+
+        protected BypassingAccessibilityBase(System.Resources.ResourceManager rspecStrings) =>
+            Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, rspecStrings);
+
+        protected override void Initialize(SonarAnalysisContext context) =>
+            FieldAccessTracker.Track(
+                context,
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/BypassingAccessibilityBase.cs:44]: I like a lot these refactorings but I think we should do them in separate PRs if they are not related to the current issue. Having a single topic for a PR makes both the review and troubleshooting (checking the history for the reason the code has changed) easier.

File: analyzers/src/SonarAnalyzer.Common/Rules/BypassingAccessibilityBase.cs [hunk 3]
```diff
@@ -26,16 +28,22 @@ public abstract class BypassingAccessibilityBase<TSyntaxKind> : SonarDiagnosticA
         where TSyntaxKind : struct
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/BypassingAccessibilityBase.cs:28]: I think the code smell raised here makes sense and we should make this a concrete class with a protected constructor.

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/BypassingAccessibility.cs [hunk 4]
```diff
@@ -31,21 +30,7 @@ namespace SonarAnalyzer.Rules.VisualBasic
...
-            DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager)
-                .WithNotConfigurable();
-
-        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } =
-            ImmutableArray.Create(rule);
-
-        public BypassingAccessibility()
-            : this(AnalyzerConfiguration.Hotspot)
-        {
-        }
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Rules/BypassingAccessibility.cs:44]: I think this change makes sense since this rule was previously a hotspot and it was changed to code smell. I also think this deserves a separate PR.

**Review by costin-zaharia-sonarsource — CHANGES_REQUESTED** [review]
I think we should split this PR in 2 parts, one related to S3011 and the other to S5693.

File: analyzers/its/sources/ManuallyAddedNoncompliantIssuesVB/IntentionalFindingsVB/S5693.vb [hunk 5]
```diff
@@ -0,0 +1,56 @@
+﻿Imports Microsoft.AspNetCore.Mvc
+
+Namespace Tests.TestCases
+
+    Public Class MyController
+        Inherits Controller
+
+        <HttpPost>
+        <DisableRequestSizeLimit()> ' Noncompliant (S5693) ^10#25 {{Make sure the content length limit is safe here.}}
```
[the remark this answers is not in the document]
Comment (csaba-sagi-sonarsource) [reply analyzers/its/sources/ManuallyAddedNoncompliantIssuesVB/IntentionalFindingsVB/S5693.vb:9]: yes, that is correct.

File: analyzers/src/SonarAnalyzer.Common/Rules/BypassingAccessibilityBase.cs (hunk 3 above)
Comment (csaba-sagi-sonarsource) [reply analyzers/src/SonarAnalyzer.Common/Rules/BypassingAccessibilityBase.cs]: It dissapeared with Pavels tracker refactoring.

File: analyzers/its/sources/ManuallyAddedNoncompliantIssuesVB/IntentionalFindingsVB/S5693.vb [hunk 6]
```diff
@@ -0,0 +1,56 @@
+﻿Imports Microsoft.AspNetCore.Mvc
```
[the remark this answers is not in the document]
Comment (csaba-sagi-sonarsource) [reply analyzers/its/sources/ManuallyAddedNoncompliantIssuesVB/IntentionalFindingsVB/S5693.vb:1]: No Core web app for VB.Net.

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/RequestsWithExcessiveLength_CustomValues.vb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+﻿Imports System
+Imports Microsoft.AspNetCore.Mvc
+
+Namespace Tests.TestCases
+
+    Public Class MyController
+        Inherits Controller
+
+        <HttpPost>
+        <RequestSizeLimit(16)> ' Noncompliant ^10#20 {{Make sure the content length limit is safe here.}}
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/RequestsWithExcessiveLength_CustomValues.vb:10]: Can you please add also a compliant example to be sure we don't raise on all?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/RequestsWithExcessiveLength_CustomValues.cs [hunk 8]
```diff
@@ -0,0 +1,24 @@
+﻿using System;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Tests.Diagnostics
+{
+    public class MyController : Controller
+    {
+        [HttpPost]
+        [RequestSizeLimit(16)] // Noncompliant {{Make sure the content length limit is safe here.}}
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/RequestsWithExcessiveLength_CustomValues.cs:9]: Can you please add also a compliant example to be sure we don't raise on all?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/RequestsWithExcessiveLength.cs [hunk 9]
```diff
@@ -0,0 +1,99 @@
+﻿using System;
+using Microsoft.AspNetCore.Mvc;
+
+namespace Tests.Diagnostics
+{
+    public class MyController : Controller
+    {
+        [HttpPost]
+        [DisableRequestSizeLimit] // Noncompliant {{Make sure the content length limit is safe here.}}
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/RequestsWithExcessiveLength.cs:9]: All these potentially vulnerable attributes can be used also at the class level. Can we add a test case also for that?

File: analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/RequestsWithExcessiveLength.cs [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+//       ^^^^^^^^^^^^^^^^^^^^^^^
+        public ActionResult PostRequestWithNoLimit()
+        {
+            return null;
+        }
+
+        [HttpPost]
+        [RequestSizeLimit(10000000)] // Noncompliant {{Make sure the content length limit is safe here.}}
```
Comment (costin-zaharia-sonarsource) [inline analyzers/tests/SonarAnalyzer.UnitTest/TestCases/Hotspots/RequestsWithExcessiveLength.cs:17]: It would be nice to use the exact boundaries 2000000 and 2000001

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/RequestsWithExcessiveLength.cs [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+    [Rule(DiagnosticId)]
+    public sealed class RequestsWithExcessiveLength : RequestsWithExcessiveLengthBase<SyntaxKind, AttributeSyntax>
+    {
+        protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
+
+        public RequestsWithExcessiveLength() : this(AnalyzerConfiguration.Hotspot) { }
+
+        internal RequestsWithExcessiveLength(IAnalyzerConfiguration analyzerConfiguration) : base(RspecStrings.ResourceManager, analyzerConfiguration) { }
+
+        protected override bool IsInvalidRequestFormLimitsAttribut(AttributeSyntax attribute, SemanticModel semanticModel) =>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/RequestsWithExcessiveLength.cs:41]: typo: `Attribute` instead of `Attribut`. Additionally we might even drop the suffix since the question (`IsInvalidReq...`) is asked on an attribute syntax parameter.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/RequestsWithExcessiveLength.cs [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+            attribute.IsKnownType(KnownType.Microsoft_AspNetCore_Mvc_RequestFormLimitsAttribute, semanticModel)
+            && attribute.ArgumentList != null
+            && attribute.ArgumentList.Arguments.FirstOrDefault(arg => FilterArgumentsOfInterest(arg)) is { } firstArgument
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/RequestsWithExcessiveLength.cs:44]: Can merge these 2 lines using `attribute.ArgumentList?...` 
and this can be simplified further: `&& attribute.ArgumentList?.Arguments.FirstOrDefault(FilterArgumentsOfInterest) is { } firstArgument`

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/RequestsWithExcessiveLength.cs [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+            && intValue > FileUploadSizeLimit;
+
+        protected override bool IsInvalidRequestSizeLimitAttribute(AttributeSyntax attribute, SemanticModel semanticModel) =>
+            attribute.IsKnownType(KnownType.Microsoft_AspNetCore_Mvc_RequestSizeLimitAttribute, semanticModel)
+            && attribute.ArgumentList != null
+            && attribute.ArgumentList.Arguments.FirstOrDefault() is { } firstArgument
+            && Language.ExpressionNumericConverter.TryGetConstantIntValue(firstArgument.Expression, out var intValue)
+            && intValue > StandardSizeLimit;
+
+        private static bool FilterArgumentsOfInterest(AttributeArgumentSyntax argument) =>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/RequestsWithExcessiveLength.cs:55]: I would rename this to `IsMultipartBodyLengthLimit`.

`OfInterest` does not add helpful information and can be dropped.

File: analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/RequestsWithExcessiveLength.cs [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+            attribute.IsKnownType(KnownType.Microsoft_AspNetCore_Mvc_RequestFormLimitsAttribute, semanticModel)
+            && attribute.ArgumentList != null
+            && attribute.ArgumentList.Arguments.FirstOrDefault(arg => FilterArgumentsOfInterest(arg)) is { } firstArgument
+            && Language.ExpressionNumericConverter.TryGetConstantIntValue(firstArgument.Expression, out var intValue)
+            && intValue > FileUploadSizeLimit;
+
+        protected override bool IsInvalidRequestSizeLimitAttribute(AttributeSyntax attribute, SemanticModel semanticModel) =>
+            attribute.IsKnownType(KnownType.Microsoft_AspNetCore_Mvc_RequestSizeLimitAttribute, semanticModel)
+            && attribute.ArgumentList != null
+            && attribute.ArgumentList.Arguments.FirstOrDefault() is { } firstArgument
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.CSharp/Rules/Hotspots/RequestsWithExcessiveLength.cs:51]: The lines can be merged by using `?.`

File: analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/RequestsWithExcessiveLengthBase.cs [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        [RuleParameter("standardSizeLimit", PropertyType.Integer, "The maximum size of regular HTTP requests (in bytes).", DefaultStandardSizeLimit)]
+        public int StandardSizeLimit { get; set; } = DefaultStandardSizeLimit;
+
+        [RuleParameter("fileUploadSizeLimit", PropertyType.Integer, "The maximum size of HTTP requests handling file uploads (in bytes).", DefaultFileUploadSizeLimit)]
+        public int FileUploadSizeLimit { get; set; } = DefaultFileUploadSizeLimit;
+
+        protected DiagnosticDescriptor Rule { get; }
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.Common/Rules/Hotspots/RequestsWithExcessiveLengthBase.cs:54]: this can be made private

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/Hotspots/RequestsWithExcessiveLength.cs [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+using Microsoft.CodeAnalysis.VisualBasic;
+using Microsoft.CodeAnalysis.VisualBasic.Syntax;
+using SonarAnalyzer.Common;
+using SonarAnalyzer.Helpers;
+
+namespace SonarAnalyzer.Rules.VisualBasic
+{
+    [DiagnosticAnalyzer(LanguageNames.VisualBasic)]
+    [Rule(DiagnosticId)]
+    public sealed class RequestsWithExcessiveLength : RequestsWithExcessiveLengthBase<SyntaxKind, AttributeSyntax>
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Rules/Hotspots/RequestsWithExcessiveLength.cs:33]: There are a few comments from C# rule which apply also here.

File: analyzers/src/SonarAnalyzer.VisualBasic/Rules/Hotspots/RequestsWithExcessiveLength.cs [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+            attribute.IsKnownType(KnownType.Microsoft_AspNetCore_Mvc_RequestSizeLimitAttribute, semanticModel)
+            && attribute.ArgumentList != null
+            && attribute.ArgumentList.Arguments.FirstOrDefault() is { } firstArgument
+            && Language.ExpressionNumericConverter.TryGetConstantIntValue(firstArgument.GetExpression(), out var intValue)
+            && intValue > StandardSizeLimit;
+
+        private static bool FilterArgumentsOfInterest(ArgumentSyntax argument) =>
+            argument is SimpleArgumentSyntax simpleArgument
+            && simpleArgument.NameColonEquals is { } nameColonEquals
+            && nameColonEquals.Name.Identifier.ValueText.Equals(MultipartBodyLengthLimit, System.StringComparison.OrdinalIgnoreCase);
```
Comment (costin-zaharia-sonarsource) [inline analyzers/src/SonarAnalyzer.VisualBasic/Rules/Hotspots/RequestsWithExcessiveLength.cs:58]:
```suggestion
            argument is SimpleArgumentSyntax { NameColonEquals: { } nameColonEquals }
            && nameColonEquals.Name.Identifier.ValueText.Equals(MultipartBodyLengthLimit, System.StringComparison.OrdinalIgnoreCase);
```

**Review by costin-zaharia-sonarsource — APPROVED** [review]
LGTM! 🚀
